File: hello-spark-scala-sbt/src/main/scala/scala/SparkDemo.scala
package com.hello.spark.scala

import org.apache.spark.{SparkConf, SparkContext}

object SparkDemo {

  def main(args: Array[String]): Unit = {

    val conf = new SparkConf()

    conf.setMaster("local").setAppName("SparkDemo-Scala")

    val context = new SparkContext(conf)

    // 创建数组
    val strArr : Array[String]= Array("hello world", "hello java", "hello hadoop", "world is good")

    val intArr : Array[Int]= Array(11, 22, 33, 44, 55)


    // 创建集合
    // List一但创建，其值不能被改变
    val left = List(1,2,3)// 方式一
    val right = List.apply(4,5,6)// 方式二
    // List连接操作
    /**
      * 集合间的链接交互
      */


    //以下操作等价
    println(left ++ right)   // List(1,2,3,4,5,6)
    println(left ++: right)  // List(1,2,3,4,5,6)
    println(right.++:(left))    // List(1,2,3,4,5,6)
    println(right.:::(left))  // List(1,2,3,4,5,6)

    //以下操作等价
    println(0 +: left)    //List(0,1,2,3)
    println(left.+:(0))   //List(0,1,2,3)
    //以下操作等价
    println(left :+ 4 )   //List(1,2,3,4)
    println(left.:+(4) )  //List(1,2,3,4)

    //以下操作等价
    println(0 :: left )     //List(0,1,2,3)
    println(left.::(0) )    //List(0,1,2,3)


    val intList : List[Int] = left ::: right

    // 打印
    println(intList(0))
    // 遍历
    intList.foreach((i : Int ) => {
      print(i)
    })
    println
    // 简写
    intList.foreach(print(_))
    println
    intList.foreach(print)
    println

    // 将集合转换成RDD
    val value = context.makeRDD(intList,1)// 方式一

    val value1 = context.parallelize(intList)// 方式二

    // 关闭资源
    context.stop()

  }

}

File: hello-spark/src/com/hello/spark/java/SparkDemo.java
package com.hello.spark.java;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SparkDemo {

	public static void main(String[] args) {
		SparkConf sparkConf = new SparkConf();

		sparkConf.setMaster("local").setAppName("SparkDemo-Java");

		JavaSparkContext sparkContext = new JavaSparkContext(sparkConf);

		// 创建RDD 方式一：SparkContext.textFile(xxx,numPartitions)
		JavaRDD<String> stringJavaRDD = sparkContext.textFile("C:\\Users\\dafochaodong\\Desktop\\wc.txt");


		// 创建数组
		String[] strArray = {"hello world", "hello java", "hello hadoop", "world is good"};

		Integer[] intArray = new Integer[]{11,22,33,44};

		// 创建集合
		List<String> strList = Arrays.asList(strArray);

		ArrayList<Integer> intList = new ArrayList<Integer>();
		intList.add(1);
		intList.add(2);
		intList.add(3);
		intList.add(4);

		// 创建RDD 方式二：SparkContext.parallelize(seq[],numPartitions)
		// 将集合转换成RDD
		JavaRDD<Integer> javaRDD = sparkContext.parallelize(intList);// 只有这一种方式


		// 创建RDD 方式三：SparkContext.parallelizePairs(seq[Tuple2],numPartitions)
		// TODO ????????????????????????????????????????
		// List<Tuple2> tuple2List = new ArrayList<Tuple2>(); List<Tuple2> 会报错 ？？？？？
		List coursesList = new ArrayList<Tuple2>();
		coursesList.add(new Tuple2<Integer, String>(1, "Spark"));
		coursesList.add(new Tuple2<Integer, String>(2, "Java"));
		coursesList.add(new Tuple2<Integer, String>(3, "Hadoop"));


		List scoresList = Arrays.asList (new Tuple2<Integer, Integer>(1, 100), new Tuple2<Integer, Integer>(2, 90), new Tuple2<Integer, Integer>(3, 70),
										new Tuple2<Integer, Integer>(1, 110), new Tuple2<Integer, Integer>(2, 95), new Tuple2<Integer, Integer>(3, 60));


		JavaPairRDD<Integer, String> courses = sparkContext.parallelizePairs(coursesList);

		JavaPairRDD<Integer, Integer> scroes = sparkContext.parallelizePairs(scoresList);

		// cogroup 算子
		JavaPairRDD<Integer, Tuple2<Iterable<String>, Iterable<Integer>>> coursesScores = courses.cogroup(scroes);

		coursesScores.foreach(new VoidFunction<Tuple2<Integer, Tuple2<Iterable<String>, Iterable<Integer>>>>() {
			/**
			 *
			 */
			private static final long serialVersionUID = 1L;
			@Override
			public void call(Tuple2<Integer, Tuple2<Iterable<String>, Iterable<Integer>>> integerTuple2Tuple2) throws Exception {
				System. out.println("Student ID:" + integerTuple2Tuple2 ._1);

				System. out.println("Student course:" + integerTuple2Tuple2 ._2 ._1);

				System. out.println("Student Score:" + integerTuple2Tuple2 ._2 ._2);

				System. out.println("------------");
			}
		});




		// 关闭资源
		sparkContext.close();

	}

}
